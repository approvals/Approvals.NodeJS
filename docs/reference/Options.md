<a id="top"></a>

# Options

<!-- toc -->
## Contents

  * [Links](#links)<!-- endToc -->

## Links

- [How to Configure Reporters with Options](../how_tos/ConfigureReportersWithOptions.md)

---

[Back to User Guide](/doc/README.md#top)
