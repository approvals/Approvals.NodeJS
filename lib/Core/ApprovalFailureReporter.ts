export interface ApprovalFailureReporter {
    name: string;

    canReportOn(): boolean;

    report(approvedFilePath: string, receivedFilePath: string): void;
}
