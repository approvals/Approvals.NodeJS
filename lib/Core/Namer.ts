export interface Namer {
  getApprovedFile(ext: string): string;
  getReceivedFile(ext: string): string;
}
