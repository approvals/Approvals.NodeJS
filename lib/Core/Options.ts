class FileOptions {
  private options: Options;
  constructor(options: Options) {
    this.options = options;
  }

  withFileExtention(extensionWithDot: string): Options {
    return this.options.modify("FileExtention", extensionWithDot);

  }

  getFileExtension(): string {
    return this.options.get("FileExtention", () => ".txt")
  }
}

export class Options {
  protected fields: { [n: string]: any };
  constructor() {
    this.fields = {}
  }

  modify(key: string, value: any): Options {
    const next = new Options();
    for (const key1 in next.fields) {
      next.fields[key1] = this.fields[key1]
    }
    next.fields[key] = value;
    return next;
  }
  forFile(): FileOptions {
    return new FileOptions(this);
  }

  get<Type>(key: string, default1: () => Type): Type {
    if (this.fields[key] == undefined) {
      return default1();
    }
    return this.fields[key]

  }
}