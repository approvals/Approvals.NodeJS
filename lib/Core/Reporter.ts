import { Config } from "../config";

export interface Reporter {
  name: string;

  canReportOn(fileName: string): boolean;

  report(
    approvedFileName: string,
    receivedFileName: string,
    options?: Partial<Config>,
  ): void;
}
