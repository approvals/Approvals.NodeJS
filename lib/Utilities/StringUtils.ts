export class StringUtils{

    static replaceAll(text: string, find: string, replace: string) {
        while (text.includes(find)){
            text = text.replace(find,replace);
        }
        return text;
    }
}
