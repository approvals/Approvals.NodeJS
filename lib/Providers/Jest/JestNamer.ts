import {expect} from "@jest/globals";
import path from "path";
import {StringUtils} from "../../Utilities/StringUtils";

const Namer: any = require("../../Namer");

export function convertToFilename(name: string): string {
    const forbidden = "\\/:?\"<>|' ";
    for (let forbiddenKey of forbidden) {
        name = StringUtils.replaceAll(name, forbiddenKey, "_");
    }
    return name;
}

export function getJestNamer(): typeof Namer {
    const state = expect.getState();
    const file = path.parse(state.testPath as string);
    const testPath = file.dir;
    const testFileName = file.name;
    const testName = convertToFilename(`${testFileName}.${state.currentTestName}`);
    return new Namer(testPath, testName);
}
