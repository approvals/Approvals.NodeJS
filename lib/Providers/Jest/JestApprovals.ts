const StringWriter = require("../../StringWriter");
const approvals = require("../../Approvals");
import { Options } from "../../Core/Options";
import { getJestNamer } from "./JestNamer";


export function verify(sut: any, options?: Options): void {
    const config = approvals.getConfig();
    config.reporters = ["diffmerge"];
    options = options || new Options()
    const writer = new StringWriter(config,  `${sut}`, options.forFile().getFileExtension());
    approvals.verifyWithControl(getJestNamer(), writer, null, config);
}

export function verifyAsJson(data: any, options?: Options): void {
    const text = JSON.stringify(data, null, "  ");
    options = options || new Options()
    options = options.forFile().withFileExtention(".json")
    verify(text, options);
}
