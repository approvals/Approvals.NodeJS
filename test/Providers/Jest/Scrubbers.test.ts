import {describe, expect, test} from "@jest/globals";
import {verify, verifyAsJson} from "../../../lib/Providers/Jest/JestApprovals";
import {convertToFilename} from "../../../lib/Providers/Jest/JestNamer";
import {Options} from "../../../lib/Core/Options";
import {StringUtils} from "../../../lib/Utilities/StringUtils";



describe("Scubbers", () => {
    test("scrub", () => {
        const text = "[1, 2, 12, 21, 121, 131, 222]";

        verify(text,new Options().withScrubber(t => StringUtils.replaceAll(t, "2", "two")));
    });

});
