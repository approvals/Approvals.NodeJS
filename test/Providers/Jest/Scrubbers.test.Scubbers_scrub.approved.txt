[1, two, 1two, two1, 1two1, 131, twotwotwo]
